server/src/main/java/org/elasticsearch/index/codec/tsdb/DocValuesConsumerUtil.java

```diff
@@ -0,0 +1,631 @@
+
+    static MergeStats compatibleWithOptimizedMerge(
+        boolean optimizedMergeEnabled,
+        FieldInfo mergeFieldInfo,
+        MergeState mergeState,
+        CheckedFunction<DocValuesProducer, FieldEntry, IOException> getEntryFunction
+    ) throws IOException {
+        if (optimizedMergeEnabled == false
+            || mergeState.needsIndexSort == false
+            || mergeFieldInfo.docValuesSkipIndexType() != DocValuesSkipIndexType.NONE) {
+            return UNSUPPORTED;
+        }
+
+        long sumNumValues = 0;
+        int sumNumDocsWithField = 0;
+
+        for (DocValuesProducer docValuesProducer : mergeState.docValuesProducers) {
+            // TODO bring back codec version check? (per field doc values producer sits between ES87TSDBDocValuesConsumer)
+            var entry = getEntryFunction.apply(docValuesProducer);
+            if (entry == null) {
+                return UNSUPPORTED;
+            }
+
+            sumNumValues += entry.numValues;
+            sumNumDocsWithField += entry.numDocsWithField;
+        }
+
+        // Documents marked as deleted should be rare. Maybe in the case of noop operation?
+        for (int i = 0; i < mergeState.liveDocs.length; i++) {
+            if (mergeState.liveDocs[i] != null) {
+                return UNSUPPORTED;
+            }
+        }
+
+        return new MergeStats(true, sumNumValues, sumNumDocsWithField);
+    }
+
+    static DocValuesProducer mergeNumericProducer(MergeStats mergeStats, FieldInfo mergeFieldInfo, MergeState mergeState) {
+        return new TsdbDocValuesProducer(mergeStats) {
+
+            @Override
+            public NumericDocValues getNumeric(FieldInfo field) throws IOException {
+                List<NumericDocValuesSub> subs = new ArrayList<>();
+                assert mergeState.docMaps.length == mergeState.docValuesProducers.length;
+                for (int i = 0; i < mergeState.docValuesProducers.length; i++) {
+                    NumericDocValues values = null;
+                    DocValuesProducer docValuesProducer = mergeState.docValuesProducers[i];
+                    if (docValuesProducer != null) {
+                        FieldInfo readerFieldInfo = mergeState.fieldInfos[i].fieldInfo(mergeFieldInfo.name);
+                        if (readerFieldInfo != null && readerFieldInfo.getDocValuesType() == DocValuesType.NUMERIC) {
+                            values = docValuesProducer.getNumeric(readerFieldInfo);
+                        }
+                    }
+                    if (values != null) {
+                        subs.add(new NumericDocValuesSub(mergeState.docMaps[i], values));
+                    }
+                }
+
+                return mergeNumericValues(subs, mergeState.needsIndexSort);
+            }
+        };
+    }
+
+    static NumericDocValues mergeNumericValues(List<NumericDocValuesSub> subs, boolean indexIsSorted) throws IOException {
```

Comment:
Did we copy these from DocValuesConsumer? Is there any chance we can avoid copying this?

Reply:
Yes. I don't see another way here. All the logic that we need here is private to `DocValuesConsumer` only.